external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala

```diff
@@ -0,0 +1,401 @@
+      batchForTime.clear()
+      generatedRDDs.foreach { kv =>
+        val a = kv._2.asInstanceOf[KafkaRDD[K, V]].offsetRanges.map(_.toTuple).toArray
+        batchForTime += kv._1 -> a
+      }
+    }
+
+    override def cleanup(time: Time) { }
+
+    override def restore() {
+      batchForTime.toSeq.sortBy(_._1)(Time.ordering).foreach { case (t, b) =>
+         logInfo(s"Restoring KafkaRDD for time $t ${b.mkString("[", ", ", "]")}")
+         generatedRDDs += t -> new KafkaRDD[K, V](
+           context.sparkContext,
+           executorKafkaParams,
+           b.map(OffsetRange(_)),
+           getPreferredHosts,
+           // during restore, it's possible same partition will be consumed from multiple
+           // threads, so dont use cache
+           false
+         )
+      }
+    }
+  }
+
+  /**
+   * A RateController to retrieve the rate from RateEstimator.
+   */
+  private[streaming] class DirectKafkaRateController(id: Int, estimator: RateEstimator)
+    extends RateController(id, estimator) {
+    override def publish(rate: Long): Unit = ()
+  }
+}
+
+/**
+ * Companion object that provides methods to create instances of [[DirectKafkaInputDStream]]
+ */
+@Experimental
+object DirectKafkaInputDStream extends Logging {
+  import org.apache.spark.streaming.api.java.{ JavaInputDStream, JavaStreamingContext }
+  import org.apache.spark.api.java.function.{ Function0 => JFunction0 }
+
+  /** Prefer to run on kafka brokers, if they are on same hosts as executors */
+  val preferBrokers: ju.Map[TopicPartition, String] = null
+  /** Prefer a consistent executor per TopicPartition, evenly from all executors */
+  val preferConsistent: ju.Map[TopicPartition, String] = ju.Collections.emptyMap()
+
+  /**
+   * Scala constructor
+   * @param preferredHosts map from TopicPartition to preferred host for processing that partition.
+   * In most cases, use [[DirectKafkaInputDStream.preferConsistent]]
+   * Use [[DirectKafkaInputDStream.preferBrokers]] if your executors are on same nodes as brokers.
+   * @param executorKafkaParams Kafka
+   * <a href="http://kafka.apache.org/documentation.html#newconsumerconfigs">
+   * configuration parameters</a>.
+   *   Requires  "bootstrap.servers" to be set with Kafka broker(s),
+   *   NOT zookeeper servers, specified in host1:port1,host2:port2 form.
+   * @param driverConsumer zero-argument function for you to construct a Kafka Consumer,
+   *  and subscribe topics or assign partitions.
+   *  This consumer will be used on the driver to query for offsets only, not messages.
+   *  See <a href="http://kafka.apache.org/documentation.html#newconsumerapi">Consumer doc</a>
+   * @tparam K type of Kafka message key
+   * @tparam V type of Kafka message value
+   */
+  def apply[K: ClassTag, V: ClassTag](
+      ssc: StreamingContext,
+      preferredHosts: ju.Map[TopicPartition, String],
+      executorKafkaParams: ju.Map[String, Object],
+      driverConsumer: () => Consumer[K, V]
```

Comment:
You need to be able to reconstruct a properly setup consumer after checkpoint recovery.  There are lots of different ways to set up a KafkaConsumer, and they require post-object-instantiation calls to e.g. subscribe or assign, with fixed or dynamic topics, etc.

Trying to do multiple overloads would get into an explosion problem pretty quickly, especially when you consider some of the quirks in the underlying API that need to be worked around for special cases.